replication-api/src/main/java/com/connexta/replication/queue/Task.java

```diff
@@ -0,0 +1,154 @@
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.queue;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Represents a task that was queued for later processing by a worker.
+ *
+ * <p>When a task is first queued, its total attempts counter will report <code>0</code> and its
+ * queued time and original queued time will be set to the same time (i.e. the current time when the
+ * task is queued via one of {@link Queue#offer(TaskInfo)}), {@link Queue#offer(TaskInfo, long,
+ * TimeUnit)}, or {@link Queue#put(TaskInfo)}. If the task fails and a decision is made to put it
+ * back into the queue via one of {@link #offerForRetry(PriorityLevel)}, {@link
+ * #offerForRetry(PriorityLevel, long, TimeUnit)}, or {@link #putForRetry(PriorityLevel)}, a new
+ * task will be created with its total attempts counter increased by <code>1</code>. The original
+ * queued time will be copied from this task and its queued time will be set to the current time
+ * when the new created task will be queued.
+ */
+public interface Task extends TaskInfo {
+  /**
+   * Gets the queue from which this task was retrieved.
+   *
+   * @return the queue from which this task was retrieved
+   */
+  public Queue getQueue();
+
+  /**
+   * Gets the time when the first attempted task was queued.
+   *
+   * @return the time when the first attempted task was queued in milliseconds from the epoch
+   */
+  public long getOriginalQueuedTime();
+
+  /**
+   * Gets the time when the task was queued. This will be the same time as reported by {@link
+   * #getOriginalQueuedTime()} if the task has not yet been attempted ({@link #getTotalAttempts()}
+   * returns <code>0</code>).
+   *
+   * @return the time when this task was queued in milliseconds from the epoch
+   */
+  public long getQueuedTime();
+
+  /**
+   * Gets the total number of times this particular task has been attempted. When a task fails for a
+   * reason that doesn't preclude it from being re-tried and it gets re-queued by the worker, its
+   * total attempts counter will automatically be incremented as it is re-added to the associated
+   * queue for later reprocessing.
+   *
+   * @return the total number of times this task has been attempted so far
+   */
+  public int getTotalAttempts();
+
+  /**
+   * Checks if this task is still locked by the current thread/worker. A task will automatically be
+   * unlocked once one of {@link #unlock()}, {@link #complete()}, {@link #fail()}, {@link
+   * #offerForRetry(PriorityLevel)}, {@link #offerForRetry(PriorityLevel, long, TimeUnit)}, or
+   * {@link #putForRetry(PriorityLevel)} has been called on it.
+   *
+   * @return <code>true</code> if the current thread/worker owns the lock to this task; <code>false
```

Comment:
❓ Why the `current` thread/worker? I figured this would be used by the broker, when a Task is requested, to skip locked Tasks that haven't yet been removed from the queue.

Reply:
This is part of the API used by clients. The state you are talking about is not something a client would need or care about and as such can be left hidden away as part of the broker's implementation of this interface.

Reply:
Why does a client care if a Task it has is locked. By virtue of the broker giving a client the Task, it should be locked for that client, right?

Reply:
A client might not care if properly implemented. However, a client which cannot imply the state based on its implementation shouldn't blindly call the methods below and simply catch the `IllegalStateException`. The contract of the methods below indicates you should not be calling them if you are not the owner so it becomes the responsibility of the client to ensure either by its design or manually by checking that they are the owner of a task before calling these methods.

In addition, implementation of this API will end up first checking if the caller if the owner of the task (or lock) by calling this exact method.

In the end, I am not exposing something they don't know about. The lock concept is part of this API and a client has to be aware of it.